velox/dwio/parquet/writer/Writer.cpp

```diff
@@ -138,49 +151,110 @@ Writer::Writer(
+
+    auto fields = arrowContext_->schema->fields();
+    std::vector<std::shared_ptr<arrow::ChunkedArray>> chunks;
+    for (int colIdx = 0; colIdx < fields.size(); colIdx++) {
+      auto dataType = fields.at(colIdx)->type();
+      auto chunk =
+          arrow::ChunkedArray::Make(
+              std::move(arrowContext_->stagingChunks.at(colIdx)), dataType)
+              .ValueOrDie();
+      chunks.push_back(chunk);
+    }
+    auto table = arrow::Table::Make(
+        arrowContext_->schema,
+        std::move(chunks),
+        static_cast<int64_t>(arrowContext_->stagingRows));
+    PARQUET_THROW_NOT_OK(arrowContext_->writer->WriteTable(
+        *table, static_cast<int64_t>(flushPolicy_->rowsInRowGroup())));
+    PARQUET_THROW_NOT_OK(stream_->Flush());
+    for (auto& chunk : arrowContext_->stagingChunks) {
+      chunk.clear();
+    }
+    arrowContext_->stagingRows = 0;
+    arrowContext_->stagingBytes = 0;
+  }
+}
+
+dwio::common::StripeProgress getStripeProgress(
+    uint64_t stagingRows,
+    int64_t stagingBytes) {
+  return dwio::common::StripeProgress{
+      .stripeRowCount = stagingRows, .stripeSizeEstimate = stagingBytes};
+}
+
+/**
+ * This method would cache input `ColumnarBatch` to make the size of row group
+ * big. It would flush when:
+ * - the cached numRows bigger than `rowsInRowGroup_`
+ * - the cached bytes bigger than `bytesInRowGroup_`
+ *
+ * This method assumes each input `ColumnarBatch` have same schema.
+ */
 void Writer::write(const VectorPtr& data) {
   ArrowArray array;
   ArrowSchema schema;
   exportToArrow(data, array, generalPool_.get());
   exportToArrow(data, schema);
   PARQUET_ASSIGN_OR_THROW(
       auto recordBatch, arrow::ImportRecordBatch(&array, &schema));
-  auto table = arrow::Table::Make(
-      recordBatch->schema(), recordBatch->columns(), data->size());
-  if (!arrowContext_->writer) {
-    auto arrowProperties = ::parquet::ArrowWriterProperties::Builder().build();
-    PARQUET_THROW_NOT_OK(::parquet::arrow::FileWriter::Open(
-        *recordBatch->schema(),
-        arrow::default_memory_pool(),
-        stream_,
-        arrowContext_->properties,
-        arrowProperties,
-        &arrowContext_->writer));
+  if (!arrowContext_->schema) {
+    arrowContext_->schema = recordBatch->schema();
+    for (int colIdx = 0; colIdx < arrowContext_->schema->num_fields();
+         colIdx++) {
+      arrowContext_->stagingChunks.push_back(
+          std::vector<std::shared_ptr<arrow::Array>>());
+    }
   }
 
-  PARQUET_THROW_NOT_OK(
-      arrowContext_->writer->WriteTable(*table, rowsInRowGroup_));
+  auto bytes = data->estimateFlatSize();
+  auto numRows = data->size();
+  if (flushPolicy_->shouldFlush(getStripeProgress(
+          arrowContext_->stagingRows, arrowContext_->stagingBytes))) {
+    flush();
+  }
+
+  for (int colIdx = 0; colIdx < recordBatch->num_columns(); colIdx++) {
+    auto array = recordBatch->column(colIdx);
+    arrowContext_->stagingChunks.at(colIdx).push_back(array);
```

Comment:
Avoid shadowing local variable
```suggestion
    arrowContext_->stagingChunks.at(colIdx).push_back(recordBatch->column(colIdx));
```